# Import air data

In [2]:
import dask
import re
import dask
import numpy as np
import pandas as pd
import math
import matplotlib.pyplot as plt

In [3]:
#read & preprocess files
#reserve
df_air_reserve=pd.read_csv('~/Desktop/ml/restaurant_vistor/air_reserve.csv')
df_air_reserve['visit_date']=df_air_reserve['visit_datetime'].str.split().str[0]
df_air_reserve['visit_time']=df_air_reserve['visit_datetime'].str.split().str[1]
df_air_reserve['reserve_date']=df_air_reserve['reserve_datetime'].str.split().str[0]
df_air_reserve['reserve_time']=df_air_reserve['reserve_datetime'].str.split().str[1]
# group by store_id & visit_date
df_air_reserve_grouped=df_air_reserve.groupby(['air_store_id','visit_date']).sum()
df_air_reserve_grouped=df_air_reserve_grouped.reset_index()

#store info
df_air_store_info=pd.read_csv('~/Desktop/ml/restaurant_vistor/air_store_info.csv')

#visit
df_air_visit_data=pd.read_csv('~/Desktop/ml/restaurant_vistor/air_visit_data.csv')

#date info
df_date_info=pd.read_csv('~/Desktop/ml/restaurant_vistor/date_info.csv')

In [4]:
# join tables df_air_reserve_grouped, df_air_store_info & df_air_visit_data
df_air=df_air_reserve_grouped.merge(df_air_store_info, left_on='air_store_id', right_on='air_store_id', how='outer')
df_air=df_air.merge(df_air_visit_data, left_on=['air_store_id', 'visit_date'], right_on=['air_store_id', 'visit_date'], how='outer')

# join date_info
df_air=df_air.merge(df_date_info, left_on=['visit_date'], right_on=['calendar_date'], how='left')

df_air.drop(columns='calendar_date', inplace=True)

In [5]:
df_air.head(10)

air_store_id  visit_date  reserve_visitors  air_genre_name  \
0  air_00a91d42b08b08d9  2016-10-31               2.0  Italian/French   
1  air_00a91d42b08b08d9  2016-12-05               9.0  Italian/French   
2  air_00a91d42b08b08d9  2016-12-14              18.0  Italian/French   
3  air_00a91d42b08b08d9  2016-12-17               2.0  Italian/French   
4  air_00a91d42b08b08d9  2016-12-20               4.0  Italian/French   
5  air_00a91d42b08b08d9  2017-02-18               9.0  Italian/French   
6  air_00a91d42b08b08d9  2017-02-23              12.0  Italian/French   
7  air_00a91d42b08b08d9  2017-03-01               3.0  Italian/French   
8  air_00a91d42b08b08d9  2017-03-14               4.0  Italian/French   
9  air_00a91d42b08b08d9  2017-03-21               3.0  Italian/French   

                     air_area_name   latitude   longitude  visitors  \
0  Tōkyō-to Chiyoda-ku Kudanminami  35.694003  139.753595      26.0   
1  Tōkyō-to Chiyoda-ku Kudanminami  35.694003  139.753595      33.0   
2  Tōkyō-to Chiyoda-ku Kudanminami  35.694003  139.753595      30.0   
3  Tōkyō-to Chiyoda-ku Kudanminami  35.694003  139.753595      17.0   
4  Tōkyō-to Chiyoda-ku Kudanminami  35.694003  139.753595      37.0   
5  Tōkyō-to Chiyoda-ku Kudanminami  35.694003  139.753595      18.0   
6  Tōkyō-to Chiyoda-ku Kudanminami  35.694003  139.753595      40.0   
7  Tōkyō-to Chiyoda-ku Kudanminami  35.694003  139.753595      17.0   
8  Tōkyō-to Chiyoda-ku Kudanminami  35.694003  139.753595      28.0   
9  Tōkyō-to Chiyoda-ku Kudanminami  35.694003  139.753595      28.0   

  day_of_week  holiday_flg  
0      Monday          0.0  
1      Monday          0.0  
2   Wednesday          0.0  
3    Saturday          0.0  
4     Tuesday          0.0  
5    Saturday          0.0  
6    Thursday          0.0  
7   Wednesday          0.0  
8     Tuesday          0.0  
9     Tuesday          0.0

In [5]:
df_air.shape

(254389, 10)

# Import hpg data

In [6]:
# reserve & visit data
df_hpg_reserve=pd.read_csv('~/Desktop/ml/restaurant_vistor/hpg_reserve.csv')
df_hpg_reserve['visit_date']=df_hpg_reserve['visit_datetime'].str.split().str[0]
df_hpg_reserve['visit_time']=df_hpg_reserve['visit_datetime'].str.split().str[1]
df_hpg_reserve['reserve_date']=df_hpg_reserve['reserve_datetime'].str.split().str[0]
df_hpg_reserve['reserve_time']=df_hpg_reserve['reserve_datetime'].str.split().str[1]
# group by store_id & visit_date
df_hpg_reserve_grouped=df_hpg_reserve.groupby(['hpg_store_id', 'visit_date']).sum()
df_hpg_reserve_grouped=df_hpg_reserve_grouped.reset_index()

#store info
df_hpg_store_info=pd.read_csv('~/Desktop/ml/restaurant_vistor/hpg_store_info.csv')

# join tables df_hpg_reserve & df_hpg_store_info
df_hpg=df_hpg_reserve_grouped.merge(df_hpg_store_info, on='hpg_store_id', how='outer')

In [7]:
df_hpg.head(10)

hpg_store_id  visit_date  reserve_visitors hpg_genre_name  \
0  hpg_001112ef76b9802c  2016-02-26                 9            NaN   
1  hpg_001112ef76b9802c  2016-03-17                 3            NaN   
2  hpg_001112ef76b9802c  2016-03-31                 5            NaN   
3  hpg_001112ef76b9802c  2016-04-05                13            NaN   
4  hpg_001112ef76b9802c  2016-04-18                 9            NaN   
5  hpg_001112ef76b9802c  2016-07-05                 3            NaN   
6  hpg_001112ef76b9802c  2016-07-14                 5            NaN   
7  hpg_001112ef76b9802c  2016-07-21                 3            NaN   
8  hpg_001112ef76b9802c  2016-07-26                 3            NaN   
9  hpg_001112ef76b9802c  2016-07-28                 3            NaN   

  hpg_area_name  latitude  longitude  
0           NaN       NaN        NaN  
1           NaN       NaN        NaN  
2           NaN       NaN        NaN  
3           NaN       NaN        NaN  
4           NaN       NaN        NaN  
5           NaN       NaN        NaN  
6           NaN       NaN        NaN  
7           NaN       NaN        NaN  
8           NaN       NaN        NaN  
9           NaN       NaN        NaN

In [8]:
df_hpg.shape

(1355053, 7)

# Join air and hpg data

In [9]:
# check number of stores in each dataframe
print(df_air['air_store_id'].nunique(), df_hpg['hpg_store_id'].nunique())

829 13325


In [92]:
df_store_id_relation=pd.read_csv('~/Desktop/ml/restaurant_vistor/store_id_relation.csv')

In [93]:
df_store_id_relation.head(10)

air_store_id          hpg_store_id
0  air_63b13c56b7201bd9  hpg_4bc649e72e2a239a
1  air_a24bf50c3e90d583  hpg_c34b496d0305a809
2  air_c7f78b4f3cba33ff  hpg_cd8ae0d9bbd58ff9
3  air_947eb2cae4f3e8f2  hpg_de24ea49dc25d6b8
4  air_965b2e0cf4119003  hpg_653238a84804d8e7
5  air_a38f25e3399d1b25  hpg_50378da9ffb9b6cd
6  air_3c938075889fc059  hpg_349b1b92f98b175e
7  air_68301bcb11e2f389  hpg_2c09f3abb2220659
8  air_5f6fa1b897fe80d5  hpg_40aff6385800ebb1
9  air_00a91d42b08b08d9  hpg_fbe603376b5980fc

In [96]:
df_store_id_relation.shape

(150, 2)

In [101]:
df_air['air_store_id'].to_frame().merge(df_store_id_relation['air_store_id'].to_frame(), on='air_store_id', how='right').nunique()

air_store_id    150
dtype: int64

In [102]:
df_hpg['hpg_store_id'].to_frame().merge(df_store_id_relation['hpg_store_id'].to_frame(), on='hpg_store_id', how='right').nunique()

hpg_store_id    150
dtype: int64

In [114]:
df=df_air.merge(df_store_id_relation, on='air_store_id', how='left')
df=df.merge(df_hpg, on=['hpg_store_id', 'visit_date'], how='left')

In [109]:
df.shape

(254389, 18)

In [116]:
df.sample(10)

air_store_id  visit_date  reserve_visitors_x  air_genre_name  \
115485  air_709262d948dd0b6e  2017-01-29                 NaN             NaN   
253994  air_c7d30ab0e07f31d5  2016-10-19                 NaN             NaN   
22938   air_bedd35489e666605  2017-04-20                 2.0    Western food   
201579  air_83db5aff8f50478e  2016-09-23                 NaN             NaN   
11175   air_55c3627912b9c849  2017-03-13                14.0  Italian/French   
171813  air_d186b2cb0b9ce022  2017-02-04                 NaN             NaN   
130950  air_fcd4492c83f1c6b9  2016-11-13                 NaN             NaN   
250806  air_7d65049f9d275c0d  2016-08-05                 NaN             NaN   
193799  air_138ee734ac79ff90  2016-02-08                 NaN             NaN   
149400  air_e89735e80d614a7e  2017-03-27                 NaN             NaN   

                                air_area_name  latitude_x  longitude_x  \
115485                                    NaN         NaN          NaN   
253994                                    NaN         NaN          NaN   
22938                Tōkyō-to Chūō-ku Tsukiji   35.670651   139.771861   
201579                                    NaN         NaN          NaN   
11175   Fukuoka-ken Fukuoka-shi Hakata Ekimae   33.591358   130.414878   
171813                                    NaN         NaN          NaN   
130950                                    NaN         NaN          NaN   
250806                                    NaN         NaN          NaN   
193799                                    NaN         NaN          NaN   
149400                                    NaN         NaN          NaN   

        visitors day_of_week  holiday_flg          hpg_store_id  \
115485      42.0      Sunday          0.0                   NaN   
253994      21.0   Wednesday          0.0                   NaN   
22938       47.0    Thursday          0.0  hpg_8f5331198ed1781d   
201579       3.0      Friday          0.0                   NaN   
11175       15.0      Monday          0.0  hpg_9d140ea335b9efd4   
171813      12.0    Saturday          0.0                   NaN   
130950      41.0      Sunday          0.0  hpg_218a1ea95def8ea3   
250806       6.0      Friday          0.0                   NaN   
193799       9.0      Monday          0.0                   NaN   
149400      23.0      Monday          0.0  hpg_b0cbb707b862dde7   

        reserve_visitors_y hpg_genre_name hpg_area_name  latitude_y  \
115485                 NaN            NaN           NaN         NaN   
253994                 NaN            NaN           NaN         NaN   
22938                  NaN            NaN           NaN         NaN   
201579                 NaN            NaN           NaN         NaN   
11175                  NaN            NaN           NaN         NaN   
171813                 NaN            NaN           NaN         NaN   
130950                 NaN            NaN           NaN         NaN   
250806                 NaN            NaN           NaN         NaN   
193799                 NaN            NaN           NaN         NaN   
149400                 2.0            NaN           NaN         NaN   

        longitude_y  
115485          NaN  
253994          NaN  
22938           NaN  
201579          NaN  
11175           NaN  
171813          NaN  
130950          NaN  
250806          NaN  
193799          NaN  
149400          NaN

# Compare air & hpg & submission

### submission data

In [103]:
df_submission=pd.read_csv('~/Desktop/ml/restaurant_vistor/sample_submission.csv')

In [104]:
df_submission.shape

(32019, 2)

In [42]:
df_submission.head(10)

id  visitors
0  air_00a91d42b08b08d9_2017-04-23         0
1  air_00a91d42b08b08d9_2017-04-24         0
2  air_00a91d42b08b08d9_2017-04-25         0
3  air_00a91d42b08b08d9_2017-04-26         0
4  air_00a91d42b08b08d9_2017-04-27         0
5  air_00a91d42b08b08d9_2017-04-28         0
6  air_00a91d42b08b08d9_2017-04-29         0
7  air_00a91d42b08b08d9_2017-04-30         0
8  air_00a91d42b08b08d9_2017-05-01         0
9  air_00a91d42b08b08d9_2017-05-02         0

In [52]:
#convert id into store_id and visit dates
df_submission['air_store_id']=df_submission['id'].str.split('_').str[1]
df_submission['air_store_id'] = 'air_'+df_submission['air_store_id']
df_submission['visit_date']=df_submission['id'].str.split('_').str[2]
df_submission.drop(columns='id', inplace=True)
df_submission.head(10)

visitors          air_store_id  visit_date
0         0  air_00a91d42b08b08d9  2017-04-23
1         0  air_00a91d42b08b08d9  2017-04-24
2         0  air_00a91d42b08b08d9  2017-04-25
3         0  air_00a91d42b08b08d9  2017-04-26
4         0  air_00a91d42b08b08d9  2017-04-27
5         0  air_00a91d42b08b08d9  2017-04-28
6         0  air_00a91d42b08b08d9  2017-04-29
7         0  air_00a91d42b08b08d9  2017-04-30
8         0  air_00a91d42b08b08d9  2017-05-01
9         0  air_00a91d42b08b08d9  2017-05-02

In [57]:
#stores in submission but not in air
set(df_submission['air_store_id'].unique())-set(df_air['air_store_id'].unique())

set()

In [50]:
df_air['air_store_id'].to_frame().merge(df_submission['air_store_id'].to_frame(), on='air_store_id', how='left').nunique()

air_store_id    829
dtype: int64

In [48]:
df_air['air_store_id'].to_frame().merge(df_submission['air_store_id'].to_frame(), on='air_store_id', how='right').nunique()

air_store_id    821
dtype: int64

#### 829 stores in air, 821 of which appears in submission data

In [56]:
#air_store_id not appeared in submission
set(df_air['air_store_id'].unique())-set(df_submission['air_store_id'].unique())

{'air_0ead98dd07e7a82a',
 'air_229d7e508d9f1b5e',
 'air_2703dcb33192b181',
 'air_b2d8bc9c88b85f96',
 'air_cb083b4789a8d3a2',
 'air_cf22e368c1a71d53',
 'air_d0a7bd3339c3d12a',
 'air_d63cfa6d6ab78446'}

In [67]:
#see if reserve_visitors are all Null
air_store_list=['air_0ead98dd07e7a82a',
 'air_229d7e508d9f1b5e',
 'air_2703dcb33192b181',
 'air_b2d8bc9c88b85f96',
 'air_cb083b4789a8d3a2',
 'air_cf22e368c1a71d53',
 'air_d0a7bd3339c3d12a',
 'air_d63cfa6d6ab78446']
df_air[(df_air['air_store_id'].isin(air_store_list)) & (df_air['reserve_visitors'].notnull())].head(10)

Empty DataFrame
Columns: [air_store_id, visit_date, reserve_visitors, air_genre_name, air_area_name, latitude, longitude, visitors, day_of_week, holiday_flg, store_pos]
Index: []

#### All stores not in submission data have NaN reserve_visitors